File: search/binary_search/recursive/python/binary_search.py
def binary_search(lst, value):
	if len(lst) == 0:
		return False
	else:
		mid = len(lst) // 2
		if lst[mid] == value:
			return True
		elif lst[mid] < value:
			return binary_search(lst[mid + 1:], value)
		else:
			return binary_search(lst[:mid], value)

def test_binary_search(lst, value, resp):
	result = binary_search(lst, value)
	assert  result == resp
	if result:
		print lst, ",has value: ", value
	else:
		print lst, ",hasn't value: ", value


test_binary_search([1,2,3,4,5,6,7,8,9], 9, True)
test_binary_search([1,2,3,4,5,6,7,8,9], 2, True)
test_binary_search([1,2,3,4,5,6,7,8,9], 11, False)
test_binary_search([1,2,3,4,5,6,7,8,9], 10, False)

File: search/binary_search/recursive/go/binary_search.go
package main

import (
	"fmt"
)

func binary_search(values []int, key int) bool {
	mid := int(len(values) / 2)

	if len(values) == 0 {
		return false
	}

	if values[mid] == key {
		return true
	} else if values[mid] > key {
		return binary_search(values[:mid], key)
	} else {
		return binary_search(values[mid+1:], key)
	}
}

func main() {
	values := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	fmt.Println(values, " has value 3? ", binary_search(values, 3))
	fmt.Println(values, " has value 9? ", binary_search(values, 9))
	fmt.Println(values, " has value 10? ", binary_search(values, 10))
	fmt.Println(values, " has value 12? ", binary_search(values, 12))
	fmt.Println(values, " has value 0? ", binary_search(values, 0))
}
